packages/keystone/plugins/addressService.js

```
@@ -0,0 +1,159 @@
+        access: readOnlyAccess,
+        hooks: {
+            ...(fieldsHooks[getSourceFieldName(mainField)] || {}),
+        },
+    },
+    [getKeyFieldName(mainField)]: {
+        type: Text,
+        schemaDoc: 'The unique key of the address',
+        access: readOnlyAccess,
+        hooks: {
+            ...(fieldsHooks[getKeyFieldName(mainField)] || {}),
+        },
+    },
+    /**
+     * @todo(nas) This field may be different for different models
+     * @see apps/condo/domains/ticket/schema/Ticket.js
+     */
+    [getMetaFieldName(mainField)]: ADDRESS_META_FIELD,
+})
+
+/**
+ * This function must return fields according to jsdoc. Returned fields are in dadata format.
+ * @param addressMeta
+ * @returns {{data: Object, unrestricted_value: String}}
+ */
+function extractAndDadatifyAddressMeta (addressMeta) {
+    const providerName = get(addressMeta, ['provider', 'name'])
+    const rawData = get(addressMeta, ['provider', 'rawData'])
+
+    let data
+    let unrestricted_value
+
+    switch (providerName) {
+        case 'dadata':
+            data = get(rawData, 'data', {})
+            unrestricted_value = String(get(rawData, 'unrestricted_value'))
+            break
+        case 'google':
+            data = {
+                // todo(nas) convert google raw data to dadata format
+            }
+            unrestricted_value = ''
+            break
+        default:
+            throw new Error(`Can not convert address meta due to no converter for '${providerName}' provider`)
+    }
+
+    return { data, unrestricted_value }
+}
+
+/**
+ * Plugin to extend address field with address service
+ * @param {string?} addressFieldName
+ * @param {Object} fieldsHooks
+ */
+const addressService = (addressFieldName = 'address', fieldsHooks = {}) => plugin(({
+    fields = {},
+    hooks = {},
+    ...rest
+}, schemaObj) => {
+    if (!addressFieldName) {
+        return { fields, hooks, ...rest }
```

it's better to throw an error

---

Done